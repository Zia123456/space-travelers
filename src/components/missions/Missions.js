import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { joinMission } from '../../redux/missions/missions';
import fetchMissions from '../../redux/missions/api';
import '../../styles/missions.css';

const Missions = () => {
  const state = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    if (state.length === 0) {
      dispatch(fetchMissions());
    }
  }, [dispatch, state]);

  const handleJoin = (id) => {
    dispatch(joinMission(id));
  };

  return (
    <div className="container">
      <section>
        <table className="border">
          <thead>
            <tr className="border">
              <th>Mission</th>
              <th>Description</th>
              <th>Status</th>
              <th aria-label="button" />
            </tr>
          </thead>
          <tbody>
            {state.map((mission) => (
              <tr key={mission.id} className="border table-row-body">
                <td className="mission-name">{mission.name}</td>
                <td className="mission-description">{mission.description}</td>
                <td className="mission-badge">
                  <p
                    className="badge-not-member"
                    style={{ display: mission.reserved ? 'none' : 'flex' }}
                  >
                    NOT A MEMBER
                  </p>
                  <p
                    className="badge-menber"
                    style={{ display: mission.reserved ? 'flex' : 'none' }}
                  >
                    Active Member
                  </p>
                </td>
                <td className="border mission-buttons">
                  <button
                    id={mission.id}
                    className="join-button"
                    type="button"
                    onClick={() => handleJoin(mission.id)}
                    style={{ display: mission.reserved ? 'none' : 'block' }}
                  >
                    Join Mission
                  </button>
                  <button
                    id={mission.id}
                    className="leave-button"
                    type="button"
                    onClick={() => handleJoin(mission.id)}
                    style={{ display: mission.reserved ? 'block' : 'none' }}
                  >
                    Leave Mission
                  </button>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </section>
    </div>
  );
};

export default Missions;
