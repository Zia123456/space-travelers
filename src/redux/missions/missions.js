// actions
const GET_MISSIONS = 'missions/GET_MISSIONS';
const JOIN_MISSION = 'missions/JOIN_MISSION';

// reducer
const initialSatate = [];

const missionsReducer = (state = initialSatate, action) => {
  switch (action.type) {
    case GET_MISSIONS:
      return action.payload;
    case JOIN_MISSION:
      return state.map((mission) => {
        if (mission.id === action.payload) {
          return {
            ...mission,
            reserved: !mission.reserved,
          };
        }
        return mission;
      });
    default:
      return state;
  }
};

// action creator
export const getMissions = (payload) => ({
  type: GET_MISSIONS,
  payload,
});

export const joinMission = (payload) => ({
  type: JOIN_MISSION,
  payload,
});

export default missionsReducer;
