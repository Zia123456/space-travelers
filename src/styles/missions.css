th {
  text-align: left;
  font-size: 1.2rem;
  padding: 10px 5px;
}

.border,
.border th,
.border td {
  border: 1px solid hsl(0, 0%, 91%);
  border-collapse: collapse;
}

.table-row-body:nth-child(odd) {
  background-color: hsl(0, 0%, 96%);
}

.table-row-body td {
  padding: 10px 5px;
  font-size: 1.1rem;
}

.mission-name {
  font-weight: 600;
}

.mission-badge,
.mission-buttons {
  width: 11%;
}

.badge-menber,
.badge-not-member {
  font-weight: 700;
  font-size: 1rem;
  color: hsl(0, 0%, 100%);
  width: 135px;
  height: 25px;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 5px;
}

.badge-not-member {
  background-color: hsl(0, 0%, 45%);
}

.badge-menber {
  background-color: hsl(194, 50%, 50%);
}

.join-button,
.leave-button {
  background-color: transparent;
  font-size: 1.1rem;
  width: 120px;
  height: 40px;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 5px;
}

.join-button {
  border: 1px solid hsl(0, 0%, 25%);
  color: hsl(0, 0%, 25%);
}

.leave-button {
  border: 1px solid rgb(233, 27, 65);
  color: rgb(233, 27, 65);
}
