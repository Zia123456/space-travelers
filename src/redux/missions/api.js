import { createAsyncThunk } from '@reduxjs/toolkit';
import { getMissions } from './missions';

const baseUrl = 'https://api.spacexdata.com/v3/missions';
const FETCH_MISSIONS = 'missions/FETCH_MISSIONS';

const fetchMissions = createAsyncThunk(FETCH_MISSIONS, async (_, thunkAPI) => {
  const response = await fetch(baseUrl, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });
  const data = await response.json();
  const missions = data.map((mission) => ({
    id: mission.mission_id,
    name: mission.mission_name,
    description: mission.description,
  }));
  thunkAPI.dispatch(getMissions(missions));

  return missions;
});

export default fetchMissions;
