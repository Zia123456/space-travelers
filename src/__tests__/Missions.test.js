import React from 'react';
import renderer from 'react-test-renderer';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import Missions from '../components/missions/Missions';
import missions from '../redux/missions/missions';

const createTestStore = () => {
  const store = configureStore({
    reducer: {
      missions,
    },
  });

  return store;
};
export default createTestStore;

describe('My Connected React-Redux Component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = createTestStore();

    component = renderer.create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    ).toJSON();
  });

  test('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });
});
