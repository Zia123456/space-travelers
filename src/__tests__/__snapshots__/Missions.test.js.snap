// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component renders correctly 1`] = `
<div
  className="container"
>
  <section>
    <table
      className="border"
    >
      <thead>
        <tr
          className="border"
        >
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
          <th
            aria-label="button"
          />
        </tr>
      </thead>
      <tbody />
    </table>
  </section>
</div>
`;
